MVC migration topic and code 3.1 update (#18796)
dotnet/AspNetCore.Docs: aspnetcore/migration/mvc.md

Invoke the rendered views from the browser on the running ASP.NET core app by replacing the current port number with the port number used in the ASP.NET core project. For example: `https://localhost:44375/home/about`.

## Migrate static content

In ASP.NET MVC 5 and earlier, static content was hosted from the root of the web project and was intermixed with server-side files. In ASP.NET Core, static content is hosted in the *wwwroot* folder.

Copy the static content from the ASP.NET MVC *WebApp1* project to the *wwwroot* folder in the ASP.NET Core *WebApp1* project:

1. In the ASP.NET Core project, right-click the *wwwroot* folder, select **Add** > **Existing Item**.
1. In the **Add Existing Item** dialog, navigate to the ASP.NET MVC *WebApp1* project.
1. Select the *favicon.ico* file, then select **Add**, replacing the existing file.

## Migrate the layout files

Copy the ASP.NET MVC project layout files to the ASP.NET Core project:

1. In the ASP.NET Core project, right-click the **Views** folder, select **Add** > **Existing Item**.

> 1. In the **Add Existing Item** dialog, navigate to the ASP.NET MVC *WebApp1* project > *Views* directory.

Author: Using the following doc as an example which (mostly) uses "directory", I think "directory" is a better route and will go that way (unless I hear otherwise):
https://docs.microsoft.com/en-us/aspnet/core/fundamentals/static-files?view=aspnetcore-3.1#serve-static-files.

1. Replace `@Scripts.Render("~/bundles/bootstrap")` with a `<script>` element (see below).

The completed replacement markup for jQuery and Bootstrap JavaScript inclusion:

```html
<script src="https://code.jquery.com/jquery-3.3.1.min.js"></script>
<script src="https://maxcdn.bootstrapcdn.com/bootstrap/3.3.7/js/bootstrap.min.js"
    integrity="sha384-Tc5IQib027qvyjSMfHjOMaLkfuWVxZxUPnCJA7l2mCWNIpG9mGCD8wGNIcPD7Txa" crossorigin="anonymous"></script>
```

The updated *_Layout.cshtml* file is shown below:

[!code-cshtml[](mvc/samples/3.x/Views/Shared/_Layout.cshtml?highlight=7-10,40-42)]

View the site in the browser. It should now render with the expected styles in place.

## Configure bundling and minification

ASP.NET Core is compatible with several open-source bundling and minification solutions such as [WebOptimizer](https://github.com/ligershark/WebOptimizer) and other similar libraries. ASP.NET Core does not provide a native bundling and minification solution. For information about how to configure bundling and minification, see [Bundling and Minification](xref:client-side/bundling-and-minification).

## Solve HTTP 500 errors

> There are many problems that can cause an HTTP 500 error message that contains no information on the source of the problem. For example, if the *Views/_ViewImports.cshtml* file contains a namespace that doesn't exist in the project, a HTTP 500 error is generated. By default in ASP.NET Core apps, the `UseDeveloperExceptionPage` extension is added to the `IApplicationBuilder` and executed when the configuration is *Development*. This is detailed in the following code:

Reviewer: ```suggestion
There are many problems that can cause an HTTP 500 error message that contains no information on the source of the problem. For example, if the *Views/_ViewImports.cshtml* file contains a namespace that doesn't exist in the project, an HTTP 500 error is generated. By default in ASP.NET Core apps, the `UseDeveloperExceptionPage` extension is added to the `IApplicationBuilder` and executed when the environment is *Development*. This is detailed in the following code:
```

[!code-csharp[](mvc/samples/3.x/Startup.cs?highlight=17-21&name=snippet)]

ASP.NET Core converts unhandled exceptions in a web app into HTTP 500 error responses. Normally, error details aren't included in these responses to prevent disclosure of potentially sensitive information about the server. For more information, see **Using the Developer Exception Page** in [Handle errors](https://docs.microsoft.com/aspnet/core/fundamentals/error-handling?view=aspnetcore-3.1#developer-exception-page).

## Next steps

> * [Migrate Authentication and Identity to ASP.NET Core](xref:migration/identity)

Reviewer: ```suggestion
* <xref:migration/identity>
```

1. From the **File** menu, select **New** > **Project**.
1. Select **ASP.NET Web Application (.NET Framework)** and then select **Next**.
1. Name the project *WebApp1* so the namespace matches the ASP.NET Core project created in the next step. Select **Create**.
1. Select **MVC**, and then select **Create**.

## Create the ASP.NET Core project

Create a new solution with a new ASP.NET Core project to migrate to:

1. Launch a second instance of Visual Studio.
1. From the **File** menu, select **New** > **Project**.
1. Select **ASP.NET Web Core Web Application** and then select **Next**.
1. In the **Configure your new project** dialog, Name the project *WebApp1*.
1. Set the location to a different directory than the previous project to use the same project name. Using the same namespace makes it easier to copy code between the two projects. Select **Create**.
1. In the **Create a new ASP.NET Core Web Application** dialog, confirm that **.NET Core** and **ASP.NET Core 3.1** are selected. Select the **Web Application (Model-View-Controller)** project template, and select **Create**.

## Configure the ASP.NET Core site to use MVC

> As of ASP.NET Core 3.0, *.NET Framework* is no longer a supported target framework, the ASP.NET Core shared framework is used instead. The ASP.NET Core shared framework contained in the [Microsoft.AspNetCore.App metapackage](xref:fundamentals/metapackage-app), no longer requires an explicit `<PackageReference />` element in the project file. The shared framework is automatically referenced when using the `Microsoft.NET.Sdk.Web` SDK in the project file:

Reviewer: ```suggestion
In ASP.NET Core 3.0 and later projects, .NET Framework is no longer a supported target framework. Your project must target .NET Core. The ASP.NET Core shared framework, which includes MVC, is part of the .NET Core runtime installation. The shared framework is automatically referenced when using the `Microsoft.NET.Sdk.Web` SDK in the project file:
```

```xml
<Project Sdk="Microsoft.NET.Sdk.Web">
```

> For more information see [Framework reference](xref:migration/22-to-30#framework-reference).

Reviewer: ```suggestion
For more information, see [Framework reference](xref:migration/22-to-30#framework-reference).
```

In ASP.NET Core, the `Startup` class:

* Replaces *Global.asax*.
* Handles all app startup tasks.

For more information, see <xref:fundamentals/startup>.

In the ASP.NET Core project, open the *Startup.cs* file:

[!code-csharp[](mvc/samples/3.x/Startup.cs?highlight=13,30,32&name=snippet)]

ASP.NET Core apps must opt in to framework features with middleware. The previous template-generated code adds the following services and middleware:

> * The <xref:Microsoft.Extensions.DependencyInjection.MvcServiceCollectionExtensions.AddControllersWithViews*> extension method registers MVC service support for controllers, API-related features, and views. For more information on MVC service registration options, see [MVC service registration](xref:migration/22-to-30#mvc-service-registration)

Reviewer: I forgot to have you encode the asterisks.

```suggestion
* The <xref:Microsoft.Extensions.DependencyInjection.MvcServiceCollectionExtensions.AddControllersWithViews%2A> extension method registers MVC service support for controllers, API-related features, and views. For more information on MVC service registration options, see [MVC service registration](xref:migration/22-to-30#mvc-service-registration)
```

This existing configuration includes what is needed to migrate the example ASP.NET MVC project. For more information on ASP.NET Core middleware options, see <xref:fundamentals/startup>.

## Migrate controllers and views

In the ASP.NET Core project, a new empty controller class and view class would be added to serve as placeholders using the same names as the controller and view classes in any ASP.NET MVC project to migrate from.

The ASP.NET Core *WebApp1* project already includes a minimal example controller and view by the same name as the ASP.NET MVC project. So those will serve as placeholders for the ASP.NET MVC controller and views to be migrated from the ASP.NET MVC *WebApp1* project.

1. Copy the methods from the ASP.NET MVC `HomeController` to replace the new ASP.NET Core `HomeController` methods. There's no need to change the return type of the action methods. The ASP.NET MVC built-in template's controller action method return type is [ActionResult](https://msdn.microsoft.com/library/system.web.mvc.actionresult(v=vs.118).aspx); in ASP.NET Core MVC, the action methods return `IActionResult` instead. `ActionResult` implements `IActionResult`.

> 1. In the ASP.NET Core project, right-click the `Views` > `Home` directory, select `Add` > `Existing Item`.

Author: @scottaddie, I originally used bold ** like you are suggesting, however Rick pointed out they might be getting localized so I may have to avoid it and suggested using code ticks.  I checked, and things like **Add** **Views** were getting localized.  So I went to went to code ticks as Rick suggested.  Maybe this is something still up for more discussion?

> 1. In the ASP.NET Core project, right-click the `Views` > `Home` directory, select `Add` > `Existing Item`.

Reviewer: We do want UI element names to be localized. If, for example, I only have the Spanish language pack installed, here's what **Add** > **New Item** looks like in Spanish:

![image](https://user-images.githubusercontent.com/10702007/85324800-13480a00-b490-11ea-8486-f71eabad55ca.png)

> 1. In the ASP.NET Core project, right-click the `Views` > `Home` directory, select `Add` > `Existing Item`.

Author: OK thanks!  Maybe I was misunderstanding a finer point of some sort.  @Rick-Anderson, we are all in agreement on going the bold ** route for UI elements such as **Add**, **New Item**?

1. In the **Add Existing Item** dialog, navigate to the ASP.NET MVC *WebApp1* project's *Views/Home* directory.
1. Select the *About.cshtml*, *Contact.cshtml*, and *Index.cshtml* Razor view files, then select `Add`, replacing the existing files.

> For more information, see [ASP.NET Core Controllers](xref:mvc/controllers/actions) and [ASP.NET Core Views](xref:mvc/views/overview).

Reviewer: ```suggestion
For more information, see <xref:mvc/controllers/actions> and <xref:mvc/views/overview>.
```

author: ardalis
description: Learn how to get started migrating an ASP.NET MVC project to ASP.NET Core MVC.
ms.author: riande
ms.date: 06/15/2020
no-loc: [Blazor, "Identity", "Let's Encrypt", Razor, SignalR]
uid: migration/mvc
---
# Migrate from ASP.NET MVC to ASP.NET Core MVC

::: moniker range=">= aspnetcore-3.0"

> This article shows how to get started migrating an ASP.NET MVC project to [ASP.NET Core MVC](../mvc/overview.md). In the process, it highlights related changes from ASP.NET MVC.

Reviewer: ```suggestion
This article shows how to start migrating an ASP.NET MVC project to [ASP.NET Core MVC](../mvc/overview.md). In the process, it highlights related changes from ASP.NET MVC.
```

> Migrating from ASP.NET MVC is a multiple step process and this article covers:

Reviewer: ```suggestion
Migrating from ASP.NET MVC is a multi-step process. This article covers:
```

* The initial setup.
* Basic controllers and views.
* Static content.
* Client-side dependencies.

> For migrating configuration and identity code, see [Migrate configuration to ASP.NET Core](configuration.md) and [Migrate Authentication and Identity to ASP.NET Core](identity.md).

Reviewer: ```suggestion
For migrating configuration and Identity code, see [Migrate configuration to ASP.NET Core](configuration.md) and [Migrate Authentication and Identity to ASP.NET Core](identity.md).
```

1. From the **File** menu, select **New** > **Project**.
1. Select **ASP.NET Web Core Web Application** and then select **Next**.
1. In the **Configure your new project** dialog, Name the project *WebApp1*.
1. Set the location to a different directory than the previous project to use the same project name. Having the same namespace makes it easier to copy code between the two projects. Select **Create**.
1. In the **Create a new ASP.NET Core Web Application** dialog, confirm that **.NET Core** and **ASP.NET Core 3.1** are selected. Select the **Web Application (Model-View-Controller)** project template, and select **Create**.

## Configure the ASP.NET Core site to use MVC

When targeting .NET Core, the [Microsoft.AspNetCore.App metapackage](xref:fundamentals/metapackage-app) is referenced by default. This package contains packages commonly used by MVC apps. If targeting .NET Framework, package references must be listed individually in the project file. `Microsoft.AspNetCore.Mvc` is the ASP.NET Core MVC framework.

In ASP.NET Core, the `Startup` class:

* Replaces *Global.asax*.
* Handles all app startup tasks.

For more information, see <xref:fundamentals/startup>.

> * In the ASP.NET Core project, open the *Startup.cs* file.

Reviewer: It feels strange to have a list with just 1 item.

```suggestion
In the ASP.NET Core project, open the *Startup.cs* file:
```

>   [!code-csharp[](mvc/samples/3.x/Startup.cs?highlight=26,43,45)]

Reviewer: Dedent the code snippet, since the list was eliminated in the suggestion above.

```suggestion
[!code-csharp[](mvc/samples/3.x/Startup.cs?highlight=26,43,45)]
```

The ASP.NET Core runtime is modular, and ASP.NET Core 3.0 added new options for registering MVC scenarios inside *Startup.ConfigureServices*.  You must also explicitly opt in for features such as serving static files and routing.

> the previous template  generated code adds the following services and middleware:

Reviewer: ```suggestion
The previous template-generated code adds the following services and middleware:
```

> In the previous code, the **Web Application (Model-View-Controller)** project template opts in the following MVC Scenario and features:

Author: I agree it would make it slimmer and the information can be found in the startup doc, however, I think it is worth keeping this section of information here.  It is a typical rather than fringe step in migration to determine what you need to opt in for your ASP.NET MVC project based on what your previous project was using, since they need to opt in now.  I think we hit the "80% rule" in this case and shouldn't have them have to switch documents and come back.

> In the previous code, the **Web Application (Model-View-Controller)** project template opts in the following MVC Scenario and features:

Reviewer 2: This write up won't tell you what you need to migrate, a link to startup helps, but even that is not enough. It's not a scenario and feature, it's a service and middleware. The doc is migration, and the brief description doesn't really help migration.

> In the previous code, the **Web Application (Model-View-Controller)** project template opts in the following MVC Scenario and features:

Reviewer 2: I question the value of most of this doc. It's really, how to migrate the simplest possible ASP.NET app to the simplest possible ASP.NET Core app. A real app will be so much more complicated,  good links to the details are what's important.

Not only are we converting the simplest possible app, we're not even including auth.

> In the previous code, the **Web Application (Model-View-Controller)** project template opts in the following MVC Scenario and features:

Author: I was referring to removing the section, which I think works better to keep and I found it helpful.  The link you mentioned is helpful too.  I'll add that.

* The `AddControllerWithViews` top-level extension method registers MVC service support for controllers, API-related features, and views. For more information on MVC service registration options, see [MVC service registration](xref:migration/22-to-30)

> * The `UseStaticFiles` extension method adds the static file handler `Microsoft.AspNetCore.StaticFiles`. The `UseStaticFiles` extension method must be called before `UseRouting`. For more information, see [Static files](xref:fundamentals/static-files).

Reviewer: ```suggestion
* The `UseStaticFiles` extension method adds the static file handler `Microsoft.AspNetCore.StaticFiles`. The `UseStaticFiles` extension method must be called before `UseRouting`. For more information, see [Static files](xref:fundamentals/static-files).
```
```suggestion
* The <xref:Microsoft.AspNetCore.Builder.StaticFileExtensions.UseStaticFiles*> extension method adds the static file handler. The method must be called before `UseRouting`. For more information, see [Static files](xref:fundamentals/static-files).
```

> * The `UseRouting` extension method adds routing. For more information, see [Application Startup](xref:fundamentals/startup) and [Routing](xref:fundamentals/routing).

Reviewer: ```suggestion
* The <xref:Microsoft.AspNetCore.Builder.EndpointRoutingApplicationBuilderExtensions.UseRouting*> extension method adds routing. For more information, see [Application Startup](xref:fundamentals/startup) and [Routing](xref:fundamentals/routing).
```

This existing configuration includes what is needed to migrate the example ASP.NET MVC project.

> ## Migrate Controllers and views

Reviewer: Sentence casing for headings

```suggestion
## Migrate controllers and views
```

> In the ASP.NET Core project a new empty controller class and view class would be added to serve as placeholders using the same names as the controller and view classes in any ASP.NET MVC project you migrate from.

Reviewer: ```suggestion
In the ASP.NET Core project, a new empty controller class and view class would be added to serve as placeholders using the same names as the controller and view classes in any ASP.NET MVC project you migrate from.
```

The ASP.NET Core *WebApp1* project already includes a minimal example controller and view by the same name as the ASP.NET MVC project. So those will serve as placeholders for the ASP.NET MVC controller and views to be migrated from the ASP.NET MVC *WebApp1* project.  

1. Copy the methods from the ASP.NET MVC `HomeController` to replace the new ASP.NET Core `HomeController` methods. There's no need to change the return type of the action methods. The ASP.NET MVC built-in template's controller action method return type is [ActionResult](https://msdn.microsoft.com/library/system.web.mvc.actionresult(v=vs.118).aspx); in ASP.NET Core MVC, the action methods return `IActionResult` instead. `ActionResult` implements `IActionResult`.
1. In the ASP.NET Core project, right-click the **Views** > **Home** folder, select  **Add** > **Existing Item...**.

> 1. In the **Add Existing Item dialog**, navigate to the ASP.NET MVC WebApp1 project > Views > Home directory.

Reviewer: ```suggestion
1. In the **Add Existing Item dialog**, navigate to the ASP.NET MVC *WebApp1* project > *Views* > *Home* directory.
```

1. Select the *About.cshtml*, *Contact.cshtml*, and *Index.cshtml* Razor view files, then select **Add**, replacing the existing files.
1. Run the ASP.NET Core app.

See [Controllers](xref:mvc/controllers/actions) and [Views](xref:mvc/views/overview) for more information.

> 

Author: @Rick-Anderson, thanks! Were there particular tip subjects you had in mind for controllers and/or views that would fit within this topics scope?  Mapping of controllers are in UseEndpoints now.  The Async suffix is removed from controller action names by default impacting routing and link generation?  I could mention those and then point to your [2.0-to-3.0 migration doc](https://docs.microsoft.com/en-us/aspnet/core/migration/22-to-30?view=aspnetcore-3.1&tabs=visual-studio)?  That work or something else you had in mind?

> 

Reviewer 2: Maybe nothing.

## Test each method

The layout file and styles have not been migrated yet, so the rendered views only contain the content in the view files.  Also, the layout file generated links for the `About` and `Contact` views will not be available yet.

* Invoke the rendered views from the browser on the running ASP.NET core app by replacing the current port number with the port number used in the ASP.NET core project. For example: `https://localhost:44375/home/about`.

> See [ASP.NET Core Controllers](xref:mvc/controllers/actions) and [ASP.NET Core Views](xref:mvc/views/overview) for more information.

Reviewer: For sentences with this structure, Acrolinx likes to see "For more information, ...".

```suggestion
For more information, see [ASP.NET Core Controllers](xref:mvc/controllers/actions) and [ASP.NET Core Views](xref:mvc/views/overview).
```

> ## Migrate Static content

Reviewer: ```suggestion
## Migrate static content
```

> In previous versions of ASP.NET MVC, static content was hosted from the root of the web project and was intermixed with server-side files. In ASP.NET Core, static content is hosted in the **wwwroot** folder.

Reviewer 2: folders are never bold.  make it a link to https://docs.microsoft.com/en-us/aspnet/core/fundamentals/?view=aspnetcore-3.1&tabs=windows#content-root

Reviewer: +1 to Rick's comment. I recommend italicizing folder names. Bolding is useful for UI element names (e.g., **Solution Explorer**).

Author: Using this doc as an example which (mostly) uses "directory", I think directory is a better route and will go that way + italics unless I hear otherwise:
https://docs.microsoft.com/en-us/aspnet/core/fundamentals/static-files?view=aspnetcore-3.1#serve-static-files.

Reviewer 2: I prefer to code fence code `wwwroot` because it's hard coded in code. Italicised words are subject to translation, although in this case very unlikely. ***Bold*** is always wrong for folders/files.

Copy the static content from the ASP.NET MVC *WebApp1* project to the *wwwroot* folder in the ASP.NET Core *WebApp1* project:

1. In the ASP.NET Core project, right-click the **wwwroot** folder, select  **Add** > **Existing Item...**.
1. In the **Add Existing Item dialog**, navigate to the ASP.NET MVC *WebApp1* project.
1. Select the *favicon.ico* file, then select **Add**, replacing the existing file.

## Migrate the layout files

Copy the ASP.NET MVC project layout files to the ASP.NET Core project:

> 1. In the ASP.NET Core project, right-click the **Views** folder, select  **Add** > **Existing Item...**.

Reviewer: ```suggestion
1. In the ASP.NET Core project, right-click the **Views** folder, select  **Add** > **Existing Item**.
```

Author: @scottaddie, thanks. Folders use italics:  What about the following situation where we are technically referring to a UI element, but it is a folder?  Still go with italics?

"In the ASP.NET Core project, right-click the *wwwroot* folder, select  **Add** > **Existing Item...**.

Reviewer: Yep, go with italics in this case

1. In the **Add Existing Item dialog**, navigate to the ASP.NET MVC *WebApp1* project > *Views* directory.
1. Select  *_ViewStart.cshtml* file then select **Add**.

Copy the ASP.NET MVC project shared layout files to the ASP.NET Core project:

> 1. In the ASP.NET Core project, right-click the **Views** > **Shared** folder, select  **Add** > **Existing Item...**.

Reviewer: ```suggestion
1. In the ASP.NET Core project, right-click the **Views** > **Shared** folder, select **Add** > **Existing Item**.
```

> 1. In the **Add Existing Item dialog**, navigate to the ASP.NET MVC *WebApp1* project > *Views* > *Shared* directory.

Reviewer: ```suggestion
1. In the **Add Existing Item** dialog, navigate to the ASP.NET MVC *WebApp1* project > *Views* > *Shared* directory.
```

Reviewer 2: > "In the ASP.NET Core project, right-click the wwwroot folder, select Add > Existing Item....

Make sure none of those get localized. Look at localized versions.

Author: @Rick-Anderson, good catch. This text is new to this update, so I didn't have a previous localized version to verify, however, I checked around in our existing content, an **Add** > and other similar right click instruction is getting localized, a wide spread problem:  

Example:
https://github.com/dotnet/AspNetCore.Docs/blob/live/aspnetcore/includes/scaffold-identity/id-scaffold-dlg.md
Compared to:
https://github.com/dotnet/AspNetCore.Docs.zh-cn/blob/live/aspnetcore/includes/scaffold-identity/id-scaffold-dlg.md

For this one I will go back and add no loc fencing and create a new issue for the repo wide problem.

Reviewer 2: Or just code fence it.

1. Replace `@Scripts.Render("~/bundles/bootstrap")` with a `<script>` element (see below).

The completed replacement markup for jQuery and Bootstrap JavaScript inclusion:

```html
<script src="https://code.jquery.com/jquery-3.3.1.min.js"></script>
<script src="https://maxcdn.bootstrapcdn.com/bootstrap/3.3.7/js/bootstrap.min.js"
    integrity="sha384-Tc5IQib027qvyjSMfHjOMaLkfuWVxZxUPnCJA7l2mCWNIpG9mGCD8wGNIcPD7Txa" crossorigin="anonymous"></script>
```

The updated *_Layout.cshtml* file is shown below:

[!code-cshtml[](mvc/samples/3.x/Views/Shared/_Layout.cshtml?highlight=7-10,40-42)]

View the site in the browser. It should now render with the expected styles in place.

## Configure bundling and minification

ASP.NET Core is compatible with several open-source bundling and minification solutions such as [WebOptimizer](https://github.com/ligershark/WebOptimizer) and other similar libraries.  ASP.NET Core does not provide a native bundling and minification solution. For information about how to configure bundling and minification, see [Bundling and Minification](../client-side/bundling-and-minification.md).

## Solve HTTP 500 errors

> There are many problems that can cause an HTTP 500 error message that contains no information on the source of the problem. For example, if the *Views/_ViewImports.cshtml* file contains a namespace that doesn't exist in your project, you'll get a HTTP 500 error. By default in ASP.NET Core apps, the `UseDeveloperExceptionPage` extension is added to the `IApplicationBuilder` and executed when the configuration is *Development*. This is detailed in the following code:

Reviewer 2: For example, if the *Views/_ViewImports.cshtml* file contains a namespace that doesn't exist ~in your project~, the app generates an internal error resulting in an HTTP 500 error.

> [!code-csharp[](mvc/samples/3.x/Startup.cs?highlight=30-35)]

Reviewer 2: put a snippet around StartUp.Configure and just show that. They don't need to see namespaces, etc.

Reviewer 2: @scottaddie  This is so VS centric, can't you assume they're run in debugger with problems like this? Do we need this?

> ASP.NET Core converts unhandled exceptions in a web app into HTTP 500 error responses. Normally, error details aren't included in these responses to prevent disclosure of potentially sensitive information about the server. See **Using the Developer Exception Page** in [Handle errors](../fundamentals/error-handling.md) for more information.

Reviewer 2: > See **Using the Developer Exception Page** in [Handle errors](../fundamentals/error-handling.md) for more information.

See https://docs.microsoft.com/en-us/aspnet/core/fundamentals/error-handling?view=aspnetcore-3.1#developer-exception-page

`#developer-exception-page` is the key.

* Copy each of the methods from the ASP.NET MVC `HomeController` to the new `HomeController`. Note that in ASP.NET MVC, the built-in template's controller action method return type is [ActionResult](https://msdn.microsoft.com/library/system.web.mvc.actionresult(v=vs.118).aspx); in ASP.NET Core MVC, the action methods return `IActionResult` instead. `ActionResult` implements `IActionResult`, so there's no need to change the return type of your action methods.

* Copy the *About.cshtml*, *Contact.cshtml*, and *Index.cshtml* Razor view files from the ASP.NET MVC project to the ASP.NET Core project.

* Run the ASP.NET Core app and test each method. We haven't migrated the layout file or styles yet, so the rendered views only contain the content in the view files. You won't have the layout file generated links for the `About` and `Contact` views, so you'll have to invoke them from the browser (replace **4492** with the port number used in your project).

  * `http://localhost:4492/home/about`

  * `http://localhost:4492/home/contact`

![Contact page](mvc/_static/contact-page.png)

Note the lack of styling and menu items. We'll fix that in the next section.

## Static content

> In previous versions of ASP.NET MVC, static content was hosted from the root of the web project and was intermixed with server-side files. In ASP.NET Core, static content is hosted in the *wwwroot* folder. You'll want to copy the static content from your old ASP.NET MVC app to the *wwwroot* folder in your ASP.NET Core project. In this sample conversion:

Reviewer 2: In ASP.NET MVC 5 and earlier, ...

Reviewer 2: Copy the static content from the ASP.NET MVC app to the *wwwroot* folder in the ASP.NET Core project.

Much better without `you`.

Reviewer 2: @scottaddie  recommend not using old ASP.NET MVC

* Just use ASP.NET MVC
* the existing ASP.NET MVC

* Copy the *favicon.ico* file from the old MVC project to the *wwwroot* folder in the ASP.NET Core project.

> The old ASP.NET MVC project uses [Bootstrap](https://getbootstrap.com/) for its styling and stores the Bootstrap files in the *Content* and *Scripts* folders. The template, which generated the old ASP.NET MVC project, references Bootstrap in the layout file (*Views/Shared/_Layout.cshtml*). You could copy the *bootstrap.js* and *bootstrap.css* files from the ASP.NET MVC project to the *wwwroot* folder in the new project. Instead, we'll add support for Bootstrap (and other client-side libraries) using CDNs in the next section.

Reviewer 2: You could is ok, probably more clear to use it.

Rather, this document adds support ...
Avoid future tense. Really avoid we.
